src/raftstore/store/peer.rs

```diff
@@ -353,6 +358,52 @@ impl Peer {
         Ok(())
     }
 
+    pub fn step(&mut self, m: eraftpb::Message) -> Result<()> {
+        if self.is_leader() && m.get_from() != INVALID_ID {
+            self.peer_heartbeats.insert(m.get_from(), Instant::now());
+        }
+        try!(self.raft_group.step(m));
+        Ok(())
+    }
+
+    pub fn check_peers(&mut self) {
+        if !self.is_leader() {
+            self.peer_heartbeats.clear();
+            return;
+        }
+
+        let peers = self.get_store().get_region().get_peers().to_owned();
```

Comment:
if the peer is leader but peer_heartbeat is empty, you can set here only, then nothing need to do. 
So this may be more simple now, only two case:

1. leader do conf change
2. leader but peer_heartbeats is empty in check_peers